### PR TITLE
transforms: add Random Erasing for image augmentation

### PR DESCRIPTION
**Pull request by zhunzhong07**
Random Erasing randomly selects a rectangle region in an image and erases its pixels with random values. It can reduce the risk of overfitting, and improves CNN baselines in image classification, object detection and person reidentification.
'Random Erasing Data Augmentation' by Zhong et.al. 
https://arxiv.org/pdf/1708.04896.pdf

### CODE REVIEW

**Comment by daavoo** [comment]
I'm not a maintainer but I think that it would be nice if you adapt the transform to follow the functional API (https://github.com/pytorch/vision/blob/master/torchvision/transforms/functional.py) for consistence with the rest of transforms. It should be trivial to move the code where the transformation actually happens to an independent function (i.e F.erase())

**Comment by zhunzhong07** [comment]
@daavoo Thanks for your advice, I have modified it.

File: torchvision/transforms/functional.py [hunk 1]
```diff
@@ -577,3 +577,48 @@ def to_grayscale(img, num_output_channels=1):
         raise ValueError('num_output_channels should be either 1 or 3')
 
     return img
+
+
+def random_erasing(img, sl=0.02, sh=0.4, r1=0.3, value=0):
```
Comment (alykhantejani) [inline torchvision/transforms/functional.py]: we should rename this to be `erase` and pass in the randomly generated values `target_area` and `aspect_ratio` thus removing the randomness from this function.

The generation of these random parameters can be moved into `RandomErasing.get_params()` as is done for example in the [RandomCrop](https://github.com/pytorch/vision/blob/master/torchvision/transforms/transforms.py#L235) transform

**Comment by zhunzhong07** [comment]
@alykhantejani Thanks for your review, I have updated it following your advice.

File: torchvision/transforms/functional.py (hunk 1 above)
Comment (zhunzhong07) [reply torchvision/transforms/functional.py]: @alykhantejani  Thanks for your review, I have updated it following your advice.

File: torchvision/transforms/functional.py [hunk 2]
```diff
@@ -577,3 +577,38 @@ def to_grayscale(img, num_output_channels=1):
...
+        w: Width of the erased image.
+        h: Height of the erased image.
+        values: Type of erasing value.
+                If values = 0, erasing with random values,
+                else, erasing with the Imagenet mean pixel value.
+
+    Returns:
+        Erased Image.
+    """
+    imagenet_mean_value = [0.485, 0.456, 0.406]
```
Comment (fmassa) [inline torchvision/transforms/functional.py]: I don't think we should hard-code the imagenet mean in the transform, but instead let the user pass it as an argument.

File: torchvision/transforms/functional.py [hunk 3]
```diff
@@ -577,3 +577,38 @@ def to_grayscale(img, num_output_channels=1):
...
+    if img.size()[0] == 3:
```
Comment (fmassa) [inline torchvision/transforms/functional.py]: Instead of having this condition, I think it would be better to just take the size of the tensor directly. Something like
```python
img[:, y:y + h, x:x + w] = torch.rand(img.size(0), h, w)
```
?

But, thinking about it again, there is *randomness* in this function (because of `torch.rand`), and I think it might be better to let the user pass the value that he wants to use as masking as an argument. So it became a single line
```python
img[:, y:y + h, x:x + w] = value
```
And if value is a `(img.size(0), 1, 1)` tensor, then it gets broadcasted naturally to fit your needs

File: torchvision/transforms/transforms.py [hunk 4]
```diff
@@ -687,3 +687,70 @@ def __call__(self, img):
...
+    Examples:
+        >>> transform = transforms.Compose([
+        >>> transforms.RandomHorizontalFlip(),
+        >>> transforms.ToTensor(),
+        >>> transforms.Normalize((0.485, 0.456, 0.406), (0.229, 0.224, 0.225)),
+        >>> transforms.RandomErasing(),
+        >>> ])
+    """
+
+    def __init__(self, probability=0.5, sl=0.02, sh=0.4, r1=0.3, value=0):
```
Comment (fmassa) [inline torchvision/transforms/transforms.py]: As I said before, it might be better to let `value` be a general Tensor or number, that can be broadcast to the correct shape

Comment (zhunzhong07) [reply torchvision/transforms/transforms.py]: @fmassa Thanks for your advice, I have updated the code. The Random Erasing is efficient in the classification task. We achieve reasonable improvements on CIFAR-10, CIFAR-100, and Fashion-MNISt datasets with various CNN models. In further, Random Erasing can also improve the baselines in object detection and person re-identification tasks. Therefore, I believe it will be widely used in CNN training for most user.

**Comment by zhunzhong07** [comment]
Hi, I found that this random erasing is now widely used in image classification (CIFAR-10 and CIFAR-100) and person re-identification. Could you consider adding it to "transforms"? @fmassa

**Comment by zhunzhong07** [comment]
New pull request at #909
